query: Add methods to create and drop caches

ethan-readyset opened the pull request:
This commit adds methods to:
- Cache a given query
- Drop the cache for a given query
- Cache all supported queries
- Drop all caches

ethan-readyset commented:
D'oh, need to add yard docs. Gimme a few mins...

helpotters reviewed (commented):
I think for the amount of specific parameter expectations this spec covers, I believe it'll be helpful to refactor the logic of repeat code into helper methods to make it more readable. 

A mostly copy-pastable example:
```ruby 
  def expect_query_attributes(query, expected_attributes)
    expect(query.id).to eq(expected_attributes[:id])
    expect(query.text).to eq(expected_attributes[:text])
    expect(query.cache_name).to eq(expected_attributes[:cache_name])
    expect(query.supported).to eq(expected_attributes[:supported])
    expect(query.count).to eq(expected_attributes[:count])

    # Handling fallback behavior, if provided
    if expected_attributes.key?(:fallback_behavior)
      expect(query.send(:fallback_behavior)).to eq(expected_attributes[:fallback_behavior])
    end
  end
```
It'll also provide coverage for any unexpected outcomes that weren't explicitly tested for but ended up being returned anyways, such as fallback support. 

This may also bring up the topic of what we envision

helpotters commented:
Also, we may be reaching a point where feature/integration specs will provide an intuitive testing interface that will provide coverage across the stack, similar to how feature tests could make most controller specs redundant. 

I don't that'll cause any changes immediately, but it might be something for us to consider later on.

helpotters commented on spec/query_spec.rb:
```diff
@@ -62,29 +40,122 @@
+      let(:query) { cached_query_attributes }
 
       before do
         allow(Readyset).
```
I found [this usage of `shared_examples`](https://github.com/thoughtbot/shoulda-matchers/blob/3654e5ca7fb10dc84a921e7f06b7875eaf8729c3/spec/unit/shoulda/matchers/active_record/have_db_index_matcher_spec.rb#L5) in the shoulda-matcher gem. 
 
 Given the repeated pattern of `setup` and `assert` instances, would something like the following help?
```ruby
  shared_examples 'query invocation' do |method, command|
    subject { send(method) }

    before do
      allow(Readyset).to receive(:raw_query).with(command).and_return(return_value)
      subject
    end

    it "invokes \"#{command}\" on Readyset" do
      expect(Readyset).to have_received(:raw_query).with(command)
    end
  end
  ```
 
 Admittedly, it was difficult for me to read the spec and understand what was going on. Your previous refactor did help a lot though! I think the coverage is excellent with *what* we're testing, but the readability in a spec should also indicate *why*.

Here's how I interpreted the `.all_seen_but_not_cached` spec after careful reading.
 ```ruby
   describe '.all_seen_but_not_cached' do
    # what we're actually executing
    subject { Readyset::Query.all_seen_but_not_cached }
    # prepare what we think the ReadySet *server* response should be. (JSON?)
    let(:seen_but_not_cached_queries) { [seen_but_not_cached_query_attributes] }

    before do
      # we are stubbing Readyset.raw_query and its response 
    end

    it 'invokes "SHOW PROXIED QUERIES" on Readyset' do
      # expect the stub to receive the right argument 
    end
```

I feel that that spec could be summarized with `it invokes SQL_COMMAND against ReadySet`. I found this [neat StackOverflow post](https://stackoverflow.com/questions/19556296/whats-the-difference-between-include-examples-and-it-behaves-like) regarding usage of `describe`, `it`, and `context`, on top of `include_examples`. 

---
**Note**:
From what I've read so far, testing should aim for *readability*, *maintainability*, and *correctness*. I am unsure if what was said above achieves that.

Here's a [neat thoughtbot article](https://thoughtbot.com/blog/the-self-contained-test) that kind of goes against `shared_examples`. 

*A quote from the article:*
> Next time you find yourself writing the same setup over and over, think about it like you’re writing a story, a piece of prose for a future developer. That future developer might even be you, and they will thank you for it.

ethan-readyset replied:
This is _amazing_ feedback -- thank you for the great suggestions, @helpotters! I will take a look at `shared_examples` and think carefully about how I can improve upon the readability of these specs 😄

helpotters commented on spec/query_spec.rb:
```diff
@@ -62,29 +40,122 @@
+      subject
+    end
+
+    it 'invokes "DROP ALL CACHES" against Readyset' do
+      expect(Readyset).to have_received(:raw_query).with('DROP ALL CACHES')
```
So, in regards to *what's* being tested, would it be accurate to say this:
- ensure that the correct SQL command is being executed.
  - *structure*: `VERB  modifier NOUN`
  - *arguments*: `query_id`(s)
- sending the `query_id` to Readyset as a parameter. 

So, at least for the class methods, the SQL command sent is defined and we're offering an easy way of invoking said command. 

While for the instance methods, such as `#cache!`, it's also surfacing information to the user, such as potential errors, or that `#reload` ran after receiving a server response.

---
another relevant thoughtbot article:  [Testing Techniques: The I/O Table](https://thoughtbot.com/blog/testing-techniques-the-i-o-table)

ethan-readyset replied:
Yes, I think that's an accurate characterization!

helpotters reviewed (commented):
@ethan-readyset feedback has arrived in your mailbox.
Mostly conversation about the contribution in-abstract with suggestions about potential refactors. However, even if you agree with the refactor ideas, we could also just make Issues for them, such as the Error classes feedback. 

As-is, everything seems solid and it has excellent coverage. This will enable a lot of other features!

ethan-readyset commented on lib/readyset/query.rb:
```diff
@@ -103,13 +149,64 @@ def initialize(attributes)
       @count = attributes['count']
```
Yup that's correct!

ethan-readyset commented:
@helpotters Made a two big changes:
- Added FactoryBot to DRY up fixture creation
- Added some shared examples to DRY up the specs massively

Let me know what you think!

helpotters reviewed (approved):
Wow, that was a quick turnaround. It looks great! Thank you for the code-review advice. I aspire to reach your level. 

![clean_specs](https://github.com/readysettech/readyset-rails/assets/22648572/38f26794-0121-4f8f-93dc-89de40060cda)

ethan-readyset commented:
Thank you again for the incredible feedback and thorough review! 😄 🚀
